File: src/providers/zenvia/schemas.ts
import { z } from "zod";

// Contract taken from docs/vendor/zenvia-openapi-v2.json:
// components.schemas["message.base"], ["content.base"], ["content.text"], ["content.template"],
// ["content.file"], ["content.email"]
const textContentSchema = z.object({
  type: z.literal("text"),
  text: z.string(),
  encodingStrategy: z.enum(["AUTO", "MORE_CHARACTER_SUPPORT", "MORE_CHARACTERS_PER_MESSAGE"]).optional(),
  reportId: z.number().optional(),
});

const templateContentSchema = z.object({
  type: z.literal("template"),
  templateId: z.string(),
  fields: z.record(z.string()).optional(),
});

const fileContentSchema = z.object({
  type: z.literal("file"),
  fileUrl: z.string().url(),
  fileName: z.string().optional(),
  fileCaption: z.string().optional(),
});

const emailContentSchema = z.object({
  type: z.literal("email"),
  subject: z.string().min(1),
  html: z.string().max(32768).optional(),
  attachments: z
    .array(z.object({ fileUrl: z.string().url(), fileName: z.string().optional() }))
    .optional(),
});

const baseMessageFields = {
  externalId: z.string().optional(),
  from: z.string().min(1).max(64),
  to: z.string().min(1).max(64),
};

// SMS: components.schemas["message.sms"], content is text | template.
export const smsContentSchema = z.discriminatedUnion("type", [textContentSchema, templateContentSchema]);
export const createSmsMessageSchema = z.object({
  ...baseMessageFields,
  contents: z.array(smsContentSchema).min(1),
});
export type CreateSmsMessageInput = z.infer<typeof createSmsMessageSchema>;

// WhatsApp: components.schemas["message.whatsapp"], content restricted here to
// text | template | file (the common cases) - the real API also supports
// buttons/lists/products/flows/location/contacts, out of scope for now.
export const whatsappContentSchema = z.discriminatedUnion("type", [
  textContentSchema,
  templateContentSchema,
  fileContentSchema,
]);
export const createWhatsappMessageSchema = z.object({
  ...baseMessageFields,
  contents: z.array(whatsappContentSchema).min(1),
  idRef: z.string().optional(),
  contentRef: z.number().optional(),
});
export type CreateWhatsappMessageInput = z.infer<typeof createWhatsappMessageSchema>;

// Email: components.schemas["message.email"], content is email | template.
export const emailMessageContentSchema = z.discriminatedUnion("type", [
  emailContentSchema,
  templateContentSchema,
]);
export const createEmailMessageSchema = z.object({
  ...baseMessageFields,
  contents: z.array(emailMessageContentSchema).min(1),
  representative: z.object({ type: z.enum(["BOT", "HUMAN"]).optional(), name: z.string().optional() }).optional(),
});
export type CreateEmailMessageInput = z.infer<typeof createEmailMessageSchema>;

export type CreateMessageInput = { externalId?: string; from: string; to: string; contents: unknown[] };

// Contract taken from docs/vendor/zenvia-openapi-v2.json:
// components.schemas["subscription.base"], ["subscription.partial-status.message-status-subscription"]
export const createSubscriptionSchema = z.object({
  eventType: z.literal("MESSAGE_STATUS"),
  webhook: z.object({
    url: z.string().url(),
    headers: z.record(z.string()).optional(),
  }),
  criteria: z.object({
    channel: z.string(),
    direction: z.enum(["IN", "OUT", "ALL"]).optional(),
  }),
});

export type CreateSubscriptionInput = z.infer<typeof createSubscriptionSchema>;

File: internal/providers/zenvia/schemas.go
package zenvia

import "net/url"

// Contract taken from docs/vendor/zenvia-openapi-v2.json:
// components.schemas["message.base"], ["content.base"], ["content.text"],
// ["content.template"], ["content.file"], ["content.email"],
// ["subscription.base"], ["subscription.partial-status.message-status-subscription"]

var contentTypesByChannel = map[string]map[string]bool{
	"sms":      {"text": true, "template": true},
	"whatsapp": {"text": true, "template": true, "file": true},
	"email":    {"email": true, "template": true},
}

type createMessageWire struct {
	ExternalID *string          `json:"externalId"`
	From       string           `json:"from"`
	To         string           `json:"to"`
	Contents   []map[string]any `json:"contents"`
	IDRef      *string          `json:"idRef,omitempty"`
	ContentRef *float64         `json:"contentRef,omitempty"`
}

func isString(v any) (string, bool) {
	s, ok := v.(string)
	return s, ok
}

func validateContentItem(channel string, item map[string]any) bool {
	typ, ok := isString(item["type"])
	if !ok || !contentTypesByChannel[channel][typ] {
		return false
	}

	switch typ {
	case "text":
		_, ok := isString(item["text"])
		return ok
	case "template":
		_, ok := isString(item["templateId"])
		return ok
	case "file":
		fileURL, ok := isString(item["fileUrl"])
		if !ok {
			return false
		}
		_, err := url.ParseRequestURI(fileURL)
		return err == nil
	case "email":
		subject, ok := isString(item["subject"])
		return ok && subject != ""
	default:
		return false
	}
}

func (w createMessageWire) valid(channel string) bool {
	if w.From == "" || len(w.From) > 64 {
		return false
	}
	if w.To == "" || len(w.To) > 64 {
		return false
	}
	if len(w.Contents) == 0 {
		return false
	}
	for _, item := range w.Contents {
		if !validateContentItem(channel, item) {
			return false
		}
	}
	return true
}

func (w createMessageWire) toContents() []any {
	out := make([]any, len(w.Contents))
	for i, c := range w.Contents {
		out[i] = c
	}
	return out
}

type createSubscriptionWire struct {
	EventType string `json:"eventType"`
	Webhook   struct {
		URL     string            `json:"url"`
		Headers map[string]string `json:"headers"`
	} `json:"webhook"`
	Criteria struct {
		Channel   string  `json:"channel"`
		Direction *string `json:"direction"`
	} `json:"criteria"`
}

func (w createSubscriptionWire) valid() bool {
	if w.EventType != "MESSAGE_STATUS" {
		return false
	}
	if w.Webhook.URL == "" {
		return false
	}
	if _, err := url.ParseRequestURI(w.Webhook.URL); err != nil {
		return false
	}
	if w.Criteria.Channel == "" {
		return false
	}
	if w.Criteria.Direction != nil {
		d := *w.Criteria.Direction
		if d != "IN" && d != "OUT" && d != "ALL" {
			return false
		}
	}
	return true
}
